### ISSUE TITLE
Play completion closure not called in Swift

### ISSUE THREAD

**Issue by federicocappelli** [airbnb/lottie-ios#118]
```
private func playAnimation() {
        self.animationView?.play(completion: { (completed) in 
        ...
        })
    }
```

With the last version (1.0.4) the animation is performed correctly, but the completion block is never called. 

animationView is a retained variable (Swift)

`fileprivate var animationView: LOTAnimationView?`

and the entire view controller is (Obj-c)

`@property (nonatomic, strong) SplashScreenViewController *splashScreenViewController;`

Note: the bug seems incredibly intermittent O_o once every 5-6 run the closure is called.

**Comment by buba447** [comment]
Which completion isn't being called? The LOT one or the UIView Animation one?
Is this on OS X or on iOS?
Im not seeing it on my end so I need a bit more context to fix :)

**Comment by federicocappelli** [comment]
Ups, sorry, the 'self.animationView?.play' completion isn't called, the animation is there only for example.
Environment: iOS 10.2, simulator and iphone 7

**Comment by federicocappelli** [comment]
Some additional information:

The animation is initialized and presented inside a View controller, and the VC is presented:
```
self.splashVC = SplashScreenViewController()
self.splashVC?.delegate = self
self.present(self.splashVC!, animated: false, completion: nil) <---- doesn't work 
//self.showDetailViewController(self.splashVC!, sender: self) <---- works!
//self.show(self.splashVC!, sender: self) <---- doesn't work
```

The VC is correctly persisted in memory.

**Comment by federicocappelli** [comment]
update:

Presenting the view controller containing the LOTAnimationView modally with:
`self.present(self.splashVC!, animated: true, completion: nil)`
the completion block is called, with: 
`self.present(self.splashVC!, animated: false, completion: nil)`
it isn't.

**Comment by federicocappelli** [comment]
Is definitely some sort of race condition, if I set breakpoints in the LOTAnimationView in the various callers of `_callCompletionIfNecesarry` the majority of the times the completion block is called.

**Comment by federicocappelli** [comment]
I'm continuing my attempts to use lottie, still hoping for help.
Now i've created a UIView named **A** with inside the LOTAnimationView named **B** configured at INIT time. Every method is called on main thread, included .play(completion: but the completion block is never called.

Should I provide example code?

**Comment by federicocappelli** [comment]
here we go: not working Swift example: [LINK](https://dl.dropboxusercontent.com/u/710772/LottieBUGExample.zip)

The SplashScreenView is almost identical to the one used in my app, only adapted.

**Comment by albertodebortoli** [comment]
+1 here

**Comment by dandongyan** [comment]
Have you guys solved this issue? I have the same issue now

**Comment by federicocappelli** [comment]
@dandongyan nope, the lib is still broken and apparently abandoned, we solved it with a dirty workaround setting our own timer.

**Comment by dandongyan** [comment]
@federicocappelli  all right, thank you, maybe i should also user timer

**Comment by rob-keepsafe** [comment]
@federicocappelli Mind posting an example of your workaround? 

We're running into this issue where Lottie doesn't call completion blocks if the animations happen when the app is in an inactive or backgrounded state (e.g. w/ wifi control panel showing / hiding)

**Comment by federicocappelli** [comment]
@rob-keepsafe here we go:

```
public func playAnimation() {
        DispatchQueue.main.async {
            self.animationView?.play()
            //            self.animationView?.play(completion: { (finished) in //DO NOT WORK
            //                self.dismiss()
            //            })
            //Note: set a timer is really bad, but lottie has a bug and the completion block isn't called, additionally the self.animationView.animationDuration is wrong. Bug reported in https://github.com/airbnb/lottie-ios/issues/118
            _ = Timer.scheduledTimer(timeInterval: TimeInterval(2.0),
                                     target: self,
                                     selector: #selector(self.dismissAnimation),
                                     userInfo: nil,
                                     repeats: false)
        }
    }
```

**Comment by rob-keepsafe** [comment]
Ah, I see. Thanks! Unfortunately mine is an indefinitely timed spinner so I'll have to look for other ways

**Comment by buba447** [comment]
Hey everyone! We are releasing an update that fixes this very soon!

**Comment by federicocappelli** [comment]
great news @buba447 thanks!

**Comment by rob-keepsafe** [comment]
@buba447 Any chance you could point me to the fix so I can patch locally for now? We're trying to release but hitting this when users show/hide the wifi control panel

**Comment by rob-keepsafe** [comment]
As a workaround for the inactive/backgrounded app scenario, here's what we're doing for now:

- When the app resigns active, we set a `shouldAnimateWithLottie` flag to `false`, reset all Lottie animations (e.g. pause them, set animation progress to 0, and hide them) and only use UIView based animations from that point forward
- As soon as the person presses a button again (meaning the app is active / foregrounded again and we have a known state), we reset that flag and allow Lottie to do its thing

**Comment by buba447** [comment]
Should be fixed in Lottie 2.0! If you still have a problem please create a new issue!
